compiler/src/iree/compiler/Codegen/Common/GPU/GPUApplyPaddingLevel.cpp

```diff
@@ -247,21 +249,32 @@ static LogicalResult applyPaddingLevel(RewriterBase &rewriter,
       assert(succeeded(reductionDimInfo) &&
              "obtained with confirmation earlier");
       for (auto &&dimInfo : reductionDimInfo.value()) {
-        Value redDimSize = tensor::DimOp::create(
-            rewriter, paddedOp.getLoc(), dimInfo.operand, dimInfo.operandDim);
+        Value redDimSize = rewriter.createOrFold<tensor::DimOp>(
+            paddedOp.getLoc(), dimInfo.operand, dimInfo.operandDim);
         reductionDimSizes.push_back({dimInfo.loopIndex, redDimSize});
       }
 
       // Add a check within the block to see if the current iteration over the
       // loops is inside or outside the padded part of the iteration space.
       rewriter.setInsertionPoint(reduction);
       SmallVector<Value> conds;
-      for (auto &&[redDim, redDimSize] : reductionDimSizes) {
-        Value redDimIndex =
-            linalg::IndexOp::create(rewriter, paddedOp.getLoc(), redDim);
-        Value cond = arith::CmpIOp::create(rewriter, paddedOp.getLoc(),
-                                           arith::CmpIPredicate::ult,
-                                           redDimIndex, redDimSize);
+      for (auto &&[dimension, unpaddedSize] : reductionDimSizes) {
+        Value cond;
+
+        // First check if the condition 'index < unpaddedSize' is always true,
+        // and directly optimize for this case.
+        int64_t lhs = paddedLoopRanges[dimension];
+        std::optional<int64_t> rhs = getConstantIntValue(unpaddedSize);
+        if (lhs != ShapedType::kDynamic && rhs.has_value() && lhs <= rhs) {
+          cond = arith::ConstantOp::create(rewriter, paddedOp.getLoc(),
```

Comment:
nit: move `paddedOp.getLoc()` to a local variable since you query it a few times?